Update dotnet new for .NET Core 3.0/3.1 release (#14356)
dotnet/docs: docs/core/tools/dotnet-new.md

---
title: dotnet new command
description: The dotnet new command creates new .NET Core projects based on the specified template.
ms.date: 02/10/2020
---
# dotnet new

> **This article applies to: ✓** .NET Core 2.0 SDK and later versions

Author: I think so but you can decide what's better. My thinking here is that these articles are usually not updated at the same time. So updating the include hasn't been viable.

I probably missed that we changed to the green one for CLI docs. Given that this one doesn't have a meaning associated with it, I thought the black one was ok here. @gewarren?

> **This article applies to: ✓** .NET Core 2.0 SDK and later versions

Reviewer: The emoji (green) check mark is better than the unicode (black) check mark because "Read aloud" on Edge reads it.

> **This article applies to: ✓** .NET Core 2.0 SDK and later versions

Author: I get that but do we need it to read aloud in this case? It's more decorative in the CLI ref docs.

> **This article applies to: ✓** .NET Core 2.0 SDK and later versions

Reviewer: For me, it doesn't detract at all to use the green check mark. Also, then you're not giving visual readers something you're not giving visually-impaired readers.

> **This article applies to: ✓** .NET Core 2.0 SDK and later versions

Author: Ok, I've added the green check here.

## Arguments

- **`TEMPLATE`**

  The template to instantiate when the command is invoked. Each template might have specific options you can pass. For more information, see [Template options](#template-options).

>   If the `TEMPLATE` value isn't an exact match on text in the **Templates** or **Short Name** column, a substring match is performed on those two columns.

Author: I might have to change this text completely since the match is against installed templates.

  Starting with .NET Core 3.0 SDK, the CLI searches for templates in NuGet.org in the following conditions:

  - If the CLI can’t find a template match when invoking `dotnet new`, not even partial.

>   - If there’s a newer version of the template available.

Author: I'll add some info but this happens automatically when you use dotnet new.

For example:
![image](https://user-images.githubusercontent.com/12971179/74482421-4d155e00-4e69-11ea-803a-94befed1537b.png)

  The command contains a default list of templates. Use `dotnet new -l` to obtain a list of the available templates. The following table shows the templates that come pre-installed with the .NET Core SDK. The default language for the template is shown inside the brackets. Click on the short name link to see the specific template options.

> | Templates                                    | Short name                      | Language     | Tags                                  | Introduced |

Author: That's how dotnet new --help calls the template category.

| Protocol Buffer File                         | [proto](#namespace)             |              | Web/gRPC                              | 3.0        |
| dotnet gitignore file                        | `gitignore`                     |              | Config                                | 3.0        |
| global.json file                             | [globaljson](#globaljson)       |              | Config                                | 2.0        |
| NuGet Config                                 | `nugetconfig`                   |              | Config                                | 1.0        |
| dotnet local tool manifest file              | `tool-manifest`                 |              | Config                                | 3.0        |
| Web Config                                   | `webconfig`                     |              | Config                                | 1.0        |
| Solution File                                | `sln`                           |              | Solution                              | 1.0        |

## Options

- **`--dry-run`**

>   Displays a summary of what would happen if the given command were run if it would result in a template creation. Available since .NET Core 2.2 SDK.

Author: Good question! Running this, it just shows the result of what would happen if you were to run the command. @KathleenDollard @dsplaisted this is the same text shown by --help. It would be good to fix this because indeed it's confusing.

- **`--force`**

  Forces content to be generated even if it would change existing files. This is required when the output directory already contains a project.

- **`-h|--help`**

  Prints out help for the command. It can be invoked for the `dotnet new` command itself or for any template, such as `dotnet new mvc --help`.

- **`-i|--install <PATH|NUGET_ID>`**

>   Installs a source or template pack from the `PATH` or `NUGET_ID` provided. If you want to install a prerelease version of a template package, you need to specify the version in the format of `<package-name>::<package-version>`. By default, `dotnet new` passes \* for the version, which represents the last stable package version. See an example at the [Examples](#examples) section.

Reviewer 2: Is a "source" a template?

```suggestion
  Installs a source or template pack from the `PATH` or `NUGET_ID` provided. If you want to install a prerelease version of a template package, you need to specify the version in the format of `<package-name>::<package-version>`. By default, `dotnet new` passes \* for the version, which represents the latest stable package version. See an example in the [Examples](#examples) section.
```

Author: Another good question. This also comes from --help. 

I'd imagine that we could just say template here. @KathleenDollard @dsplaisted?

Reviewer 3: I'm not sure what a source pack would be.  So it could probably just say "Installs a template pack"

  
  If a version of the template was already installed when you run this command, the template will be updated to the specified version, or to the latest stable version if no version was specified.

>   For information on creating custom templates, see [Custom templates for dotnet new](custom-templates.md).

Reviewer 2: ```suggestion
  For more information, see [Custom templates for dotnet new](custom-templates.md).
```

Author: It's ok to say what you'd get more information for.

- **`-l|--list`**

  Lists templates containing the specified name. If no name is specified, lists all templates.

- **`-lang|--language {C#|F#|VB}`**

  The language of the template to create. The language accepted varies by the template (see defaults in the [arguments](#arguments) section). Not valid for some templates.

  > [!NOTE]

>   > Some shells interpret `#` as a special character. In those cases, you need to enclose the language parameter value, such as `dotnet new console -lang "F#"`.

Reviewer 2: ```suggestion
  > Some shells interpret `#` as a special character. In those cases, enclose the language parameter value in quotes. For example, `dotnet new console -lang "F#"`.
```

- **`-n|--name <OUTPUT_NAME>`**

  The name for the created output. If no name is specified, the name of the current directory is used.

- **`--nuget-source`**

  Specifies a NuGet source to use during install. Available since .NET Core 2.1 SDK.

- **`-o|--output <OUTPUT_DIRECTORY>`**

  Location to place the generated output. The default is the current directory.

- **`--type`**

  Filters templates based on available types. Predefined values are "project", "item", or "other".

- **`-u|--uninstall [PATH|NUGET_ID]`**

>   Uninstalls a source or template pack at the `PATH` or `NUGET_ID` provided. When excluding the `<PATH|NUGET_ID>` value, all currently installed template packs and their associated templates are displayed. When specifying `NUGET_ID`, only specify the package name without the version number.

Reviewer 2: "source" again

>   Uninstalls a source or template pack at the `PATH` or `NUGET_ID` provided. When excluding the `<PATH|NUGET_ID>` value, all currently installed template packs and their associated templates are displayed. When specifying `NUGET_ID`, only specify the package name without the version number.

Reviewer 2: ```suggestion
  Uninstalls a source or template pack at the `PATH` or `NUGET_ID` provided. When the `<PATH|NUGET_ID>` value isn't specified, all currently installed template packs and their associated templates are displayed. When specifying `NUGET_ID`, don't include the version number.
```

  If you don't specify a parameter to this option, the command lists the installed templates and details about them.

  > [!NOTE]
  > To uninstall a template using a `PATH`, you need to fully qualify the path. For example, *C:/Users/\<USER>/Documents/Templates/GarciaSoftware.ConsoleTemplate.CSharp* will work, but *./GarciaSoftware.ConsoleTemplate.CSharp* from the containing folder will not.

>   > Additionally, do not include a final terminating directory slash on your template path.

Reviewer 2: ```suggestion
  > Don't include a final terminating directory slash on your template path.
```

- **`--update-apply`**

  Checks if there are updates available for the template packs that are currently installed and installs them. Available since .NET Core 3.0 SDK.

- **`--update-check`**

  Checks if there are updates available for the template packs that are currently installed. Available since .NET Core 3.0 SDK.

## Template options

Each project template may have additional options available. The core templates have the following additional options:

### console

- **`-f|--framework <FRAMEWORK>`**

  Specifies the [framework](../../standard/frameworks.md) to target. Available since .NET Core 3.0 SDK.

  The following table lists the default values according to the SDK version number you're using:

  | SDK version | Default value   |
  |-------------|-----------------|
  | 3.1         | `netcoreapp3.1` |
  | 3.0         | `netcoreapp3.0` |

- **`--langVersion <VERSION_NUMBER>`**

>   Sets the `LangVersion` property in the created project file. For example, use `--langVersion 7.3` to use C# 7.3. Not supported for F#. Available since .NET Core 2.2 SDK.

Reviewer 2: What's the default language version?

Author: Good question. @KathleenDollard? @BillWagner? do you know the mapping?

Author: I found the answer. I'll just link to that.
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version#defaults
